pilot/pkg/model/jwks_resolver.go

```diff
@@ -259,24 +277,37 @@ func (r *jwksResolver) getRemoteContent(uri string) ([]byte, error) {
 		client = r.secureHTTPClient
 	}
 
-	resp, err := client.Get(uri)
-	if err != nil {
-		return nil, err
-	}
-	defer func() {
-		_ = resp.Body.Close()
-	}()
+	getPublicKey := func() ([]byte, error) {
+		resp, err := client.Get(uri)
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			_ = resp.Body.Close()
+		}()
+
+		if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+			return nil, fmt.Errorf("unsuccessful response from %q", uri)
+		}
 
-	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
-		return nil, fmt.Errorf("unsuccessful response from %q", uri)
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		return body, nil
 	}
 
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return nil, err
+	for i := 1; i < jwtPubKeyFetchRetryCount; i++ {
+		if body, err := getPublicKey(); err == nil {
+			return body, nil
+		}
+		time.Sleep(jwtPubKeyFetchRetryInterval)
```

Comment:
exp backoff ?

Reply:
Could this `getRemoteContent` be called in a critical path? I.e it doesn't run on background thread, but could be in the same thread handle xDS request. If so, having long sleep could be a trouble. We may need to have this retry logic conditionally on when the call was triggered: from background jwt refresh thread, or from config push flow.

Reply:
+1, only retry on non-critical path(in timerjob).
critical path: look up cache -> (if not found) send only one request regardless of result.

Reply:
First, we could reduce the retry count and interval, Second, the only chance this gets called in a critical path is when we have a newly applied authN policy. In this case, the jwks uri is new to us and we have to fetch it from the network and then store it in the cache, after that, the critical path will always get it from the cache.

With that said, if the network error happened in the critical path in the first time, we will have a window for not having jwt public key.

Instead of not having retry, how about to make this configurable in the argument, and only retry 1 times in the critical path, but retry 3 times in the refresh job?